src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.CoreRT.cs

```diff
@@ -151,11 +156,30 @@ public ManagedObjectWrapperHolder(ManagedObjectWrapper* wrapper)
             }
         }
 
-        internal unsafe struct IUnknownVftbl
+        internal unsafe class NativeObjectWrapper
         {
-            public delegate* unmanaged<IntPtr, ref Guid, out IntPtr, int> QueryInterface;
-            public delegate* unmanaged<IntPtr, uint> AddRef;
-            public delegate* unmanaged<IntPtr, uint> Release;
+            private readonly IntPtr _externalComObject;
+            private readonly ComWrappers _comWrappers;
+            public GCHandle _proxyHandle;
+
+            public NativeObjectWrapper(IntPtr externalComObject, ComWrappers comWrappers, object comProxy)
+            {
+                _externalComObject = externalComObject;
+                _comWrappers = comWrappers;
+                _proxyHandle = GCHandle.Alloc(comProxy, GCHandleType.Weak);
+                Marshal.AddRef(externalComObject);
+            }
+
+            ~NativeObjectWrapper()
+            {
+                _comWrappers.RemoveRCWFromCache(_externalComObject);
+                if (_proxyHandle.IsAllocated)
+                {
+                    _proxyHandle.Free();
+                }
+
+                Marshal.Release(_externalComObject);
```

Comment:
```suggestion
                if (_proxyHandle.IsAllocated)
                {
                    _proxyHandle.Free();
                    Marshal.Release(_externalComObject);
                }
```
When the handle allocation fails, we are going to miss the AddRef as well.

Alternative (and maybe easier to read?) way to fix this would be to reorder the operations in the constructor.

Reply:
I'm actually for alternative and reorder operations in the constructor.